[JSC] Improve error messages for destructuring null or undefined values

sosukesuzuki opened the pull request:
#### 616723d96e05251f39c3def2fa1790da67feec2d
<pre>
[JSC] Improve error messages for destructuring null or undefined values
<a href="https://bugs.webkit.org/show_bug.cgi?id=297773">https://bugs.webkit.org/show_bug.cgi?id=297773</a>

Reviewed by Yusuke Suzuki.

Destructuring null or undefined values is invalid:

  ({ a } = null);

JSC throws a TypeError with the message &quot;Right side of assignment cannot be destructured&quot;
for such code. However, this error is thrown not only for assignments but also for
function calls, making it not a good error message.

This patch fixes that issue.

Additionally, this patch includes the name of the first destructured variable in the
error message, similar to V8. This information is helpful for debugging.

* JSTests/stress/destructuring-assignment-require-object-coercible-params.js: Added.
(testTypeError):
(testTypeError.fn):
* JSTests/stress/destructuring-assignment-require-object-coercible-symbol-private.js: Added.
(shouldThrow):
(throw.new.Error):
* JSTests/stress/destructuring-assignment-require-object-coercible.js:
* JSTests/stress/object-rest-deconstruct.js:
(catch):
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/allow-attribute-with-create.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/protocol/vp8-fmtp-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/digital-credentials/allow-attribute-with-create.https-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/digital-credentials/allow-attribute-with-create.https-expected.txt:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitRequireObjectCoercibleForDestructuring):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ObjectPatternNode::bindValue const):

Canonical link: <a href="https://commits.webkit.org/299244@main">https://commits.webkit.org/299244@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dba14a983e7dadd932366d46a8206408af32a36d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117679 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37357 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27984 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123813 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69690 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38049 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45939 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89314 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/51768 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120631 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30290 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105514 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; run-api-tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69805 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29353 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23630 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67471 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/109789 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99707 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23810 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126906 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/116188 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44582 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33553 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97971 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44940 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101741 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97758 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24981 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43135 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21097 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40947 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44453 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50128 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144886 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43912 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37282 "Found 1 new JSC binary failure: testapi, Found 20039 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/LastUsedSegmentHasNULLElement.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47259 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45603 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -4937,6 +4937,19 @@ void BytecodeGenerator::emitRequireObjectCoercible(RegisterID* value, ASCIILiter
+    Ref<Label> target = newLabel();
+    OpJnundefinedOrNull::emit(this, value, target->bind(this));
+
+    if (propertyName && !propertyName->isNull())
```
Does this need to check if it has a public name and/or it's not a symbol?

sosukesuzuki replied:
it doesn't need check explicitly, I added tests for it ( https://github.com/WebKit/WebKit/pull/49762/files#diff-c7136627367e6232c602ca6a8fb12a1527c4982345c7f9fba40692d093a76dd3 )

sosukesuzuki replied:
Ah sorry, I misunderstood. I added `!firstTarget.propertyName.isPrivateName()` to `NodeCodegen`

kmiller68 commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -4937,6 +4937,19 @@ void BytecodeGenerator::emitRequireObjectCoercible(RegisterID* value, ASCIILiter
+    OpJnundefinedOrNull::emit(this, value, target->bind(this));
+
+    if (propertyName && !propertyName->isNull())
+        emitThrowTypeError(Identifier::fromString(m_vm, makeString("Cannot destructure property '"_s, propertyName->string(), "' as it is null or undefined"_s)));
```
I would read `"TypeError: Cannot destructure property 'a' as it is null or undefined"` as 'a' is null or undefined but that's not the error here. Maybe `"TypeError: Cannot destructure property 'a' from null or undefined value"` would be a better message.

sosukesuzuki replied:
Thank you, fixed!

darinadler commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -4937,6 +4937,19 @@ void BytecodeGenerator::emitRequireObjectCoercible(RegisterID* value, ASCIILiter
+    OpJnundefinedOrNull::emit(this, value, target->bind(this));
+
+    if (propertyName && !propertyName->isNull())
+        emitThrowTypeError(Identifier::fromString(m_vm, makeString("Cannot destructure property '"_s, propertyName->string(), "' from null or undefined value"_s)));
+    else
+        emitThrowTypeError("Cannot destructure null or undefined value"_s);
```
Slightly annoying to have the message be vague: "null or undefined". Why not have separate messages for null and undefined? Would emitting an extra branch in the throwing case be too much of a waste?

sosukesuzuki replied:
I also considered distinguishing between null and undefined checks. However, checking them separately would increase the number of bytecode instructions.

Constellation replied:
Yeah, checking each one separately is too costly since size of bytecode is affecting on performance and memory severely. For now, I think this is fine.

sosukesuzuki committed:
[JSC] Improve error messages for destructuring null or undefined values
https://bugs.webkit.org/show_bug.cgi?id=297773

Reviewed by Yusuke Suzuki.

Destructuring null or undefined values is invalid:

  ({ a } = null);

JSC throws a TypeError with the message "Right side of assignment cannot be destructured"
for such code. However, this error is thrown not only for assignments but also for
function calls, making it not a good error message.

This patch fixes that issue.

Additionally, this patch includes the name of the first destructured variable in the
error message, similar to V8. This information is helpful for debugging.

* JSTests/stress/destructuring-assignment-require-object-coercible-params.js: Added.
(testTypeError):
(testTypeError.fn):
* JSTests/stress/destructuring-assignment-require-object-coercible-symbol-private.js: Added.
(shouldThrow):
(throw.new.Error):
* JSTests/stress/destructuring-assignment-require-object-coercible.js:
* JSTests/stress/object-rest-deconstruct.js:
(catch):
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/allow-attribute-with-create.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/protocol/vp8-fmtp-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/digital-credentials/allow-attribute-with-create.https-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/digital-credentials/allow-attribute-with-create.https-expected.txt:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitRequireObjectCoercibleForDestructuring):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ObjectPatternNode::bindValue const):

Canonical link: https://commits.webkit.org/299244@main
